src/QMCSimulation.jl

```diff
@@ -0,0 +1,395 @@
+    report_metadata!(report, "message", message[])
+    report_metadata!(report, "elapsed_time", elapsed_time[])
+    return report
+end
+
+# iteration for backward compatibility
+# undocumented; may be removed in the future
+function Base.iterate(sm::QMCSimulation, state=1)
+    if state == 1
+        return DataFrame(sm), 2
+    elseif state == 2
+        return sm.qmc_state, 3
+    else
+        return nothing
+    end
+end
+# getproperty access to :df and :state fields for backward compatibility
+# undocumented; may be removed in the future
+function Base.getproperty(sm::QMCSimulation, key::Symbol)
+    if key == :df
+        return DataFrame(sm)
+    elseif key == :state
+        return sm.qmc_state
+    else
+        return getfield(sm, key)
+    end
+end
+
+# Tables.jl integration: provide access to report.data
+# Note that using the `Tables` interface will not include metadata in the output.
+# To include metadata, use the `DataFrame` constructor.
+Tables.istable(::Type{<:QMCSimulation}) = true
+Tables.columnaccess(::Type{<:QMCSimulation}) = true
+Tables.columns(sm::QMCSimulation) = Tables.columns(sm.report.data)
+Tables.schema(sm::QMCSimulation) = Tables.schema(sm.report.data)
+
+
+# TODO: interface for reading results
+
+num_replicas(s::QMCSimulation) = num_replicas(s.qmc_problem)
+DataFrames.DataFrame(s::QMCSimulation) = DataFrame(s.report)
+
+"""
+    init(problem::FCIQMCProblem; copy_vectors=true)::QMCSimulation
+
+Initialise a [`Rimu.QMCSimulation`](@ref).
+
+See also [`FCIQMCProblem`](@ref), [`solve!`](@ref), [`solve`](@ref), [`step!`](@ref),
+[`Rimu.QMCSimulation`](@ref).
+"""
+function CommonSolve.init(problem::FCIQMCProblem; copy_vectors=true)
+    return QMCSimulation(problem; copy_vectors)
+end
+
+"""
+    step!(sm::QMCSimulation)::QMCSimulation
+
+Advance the simulation by one step.
+
+Calling [`solve!`](@ref) will advance the simulation until the last step or the walltime is
+exceeded. When completing the simulation without calling [`solve!`](@ref), the simulation
+report needs to be finalised by calling [`Rimu.finalize_report!`](@ref).
+
+See also [`FCIQMCProblem`](@ref), [`init`](@ref), [`solve!`](@ref), [`solve`](@ref),
+[`Rimu.QMCSimulation`](@ref).
+"""
+function CommonSolve.step!(sm::QMCSimulation)
+    @unpack qmc_state, report, modified, aborted, success, message = sm
+    @unpack replicas, simulation_plan, step, r_strat,
+        replica = qmc_state
+
+    if aborted[] || success[]
+        @warn "Simulation is already aborted or finished."
+        return sm
+    end
+    if step[] >= simulation_plan.last_step
+        @warn "Simulation has already reached the last step."
+        return sm
+    end
+
+    step[] += 1
+
+    # report step number
+    if step[] % reporting_interval(r_strat) == 0
+        report!(r_strat, step[], report, :steps, step[])
+    end
+
+    proceed = true
+    # advance all replicas
+    for replica in replicas
+        proceed &= advance!(report, qmc_state, replica)
```

Comment:
Should we rename `advance!` to step?

Reply:
You mean overload `CommonSolve.step!` for a `QMCState`?